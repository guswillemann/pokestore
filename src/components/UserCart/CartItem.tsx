import styled from 'styled-components';
import useUserCart, { ProductType } from '../../contexts/UserCartContext';
import { getThemeColor } from '../../theme/utils';

const CartItemWrapper = styled.li`
  @keyframes slideLeft {
    0% { transform: translateX(100%) }
    100% { transform: translateX(0) }
  }

  animation: slideLeft 500ms ease-in;

  display: grid;
  gap: 1rem;
  grid-template-columns: 4rem 1fr 10rem 1rem;
  height: 5rem;
  background-color: ${getThemeColor('background')};
  padding: 0.5rem;
  border-radius: ${({ theme }) => theme.borderRadius};

  p {
    align-self: center;
  }

  .price-tag {
    margin-left: auto;
  }

  button {
    background-color: transparent;
    border: none;
    cursor: pointer;
    font-size: 0;
  }

  button:hover, button:focus {
    opacity: 0.5;
  }
`;

type CartItemPropsTypes = {
  product: ProductType;
  index: number;
}

export default function CartItem({ product, index }: CartItemPropsTypes) {
  const { removeItem } = useUserCart();

  return (
    <CartItemWrapper>
      <img src={product.img} alt="Imagem do produto" />
      <p>{product.name}</p>
      <p className="price-tag">{`R$ ${product.price}`}</p>
      <button type="button" onClick={() => removeItem(index)}>
        <img src="/images/x-icon.svg" alt="X" aria-label="Remover item do carrinho" />
      </button>
    </CartItemWrapper>
  );
}
