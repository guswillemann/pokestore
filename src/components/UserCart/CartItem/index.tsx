import useUserCart, { ProductType } from '../../../contexts/UserCartContext';
import Button from '../../Button';
import ProductImage from '../../ProductImage';
import CartItemWrapper from './styles';

type CartItemPropsTypes = {
  product: ProductType;
  index: number;
}

export default function CartItem({ product, index }: CartItemPropsTypes) {
  const { removeItem } = useUserCart();

  return (
    <CartItemWrapper data-testid="cart-item">
      <ProductImage src={product.img} />
      <div className="item-info">
        <p className="item-name">{product.name}</p>
        <p className="item-price">{`$${product.price}`}</p>
        <Button
          type="button"
          onClick={() => removeItem(index)}
          aria-label="Remove item from cart"
        >
          remove
        </Button>
      </div>
    </CartItemWrapper>
  );
}
