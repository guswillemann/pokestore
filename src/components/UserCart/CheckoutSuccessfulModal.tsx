import styled from 'styled-components';
import useModal from '../../contexts/ModalContext';
import { getThemeColor } from '../../theme/utils';

const CheckoutSuccessfulModalWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 40rem;
  height: 30rem;
  position: relative;
  
  background-color: ${getThemeColor('primary')};
  color: white;

  border-radius: ${({ theme }) => theme.borderRadius};
  
  button {
    position: absolute;
    top: 1rem;
    right: 1rem;

    width: 3rem;

    background-color: transparent;
    border: none;
    cursor: pointer;

    filter: invert();

    &:hover {
      opacity: 0.8;
    }
  }

  h2 {
    margin-bottom: 2rem;
  }

  p {
    margin-top: 2rem;
  }
`;

export default function CheckoutSuccessfulModal({ CheckoutTotalPrice }: { CheckoutTotalPrice: string }) {
  const { toggleModal } = useModal();

  return (
    <CheckoutSuccessfulModalWrapper>
      <button type="button" onClick={toggleModal}>
        <img src="/images/x-icon.svg" alt="X" aria-label="Fechar modal" />
      </button>
      <h2>Obrigado !!!</h2>
      <p>VocÃª ganhou de volta</p>
      <p>{`R$ ${CheckoutTotalPrice}`}</p>
    </CheckoutSuccessfulModalWrapper>
  );
}