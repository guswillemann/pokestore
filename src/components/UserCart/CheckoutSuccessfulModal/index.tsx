import useModal from '../../../contexts/ModalContext';
import CheckoutSuccessfulModalWrapper from './styles';

export default function CheckoutSuccessfulModal({ CheckoutTotalPrice }: { CheckoutTotalPrice: string }) {
  const { toggleModal } = useModal();

  return (
    <CheckoutSuccessfulModalWrapper data-testid="checkout-modal">
      <button type="button" onClick={toggleModal}>
        <img src="/images/x-icon.svg" alt="X" aria-label="Close modal" />
      </button>
      <h2>Thank you !!!</h2>
      <p>Total price for your order:</p>
      <p>{`$${CheckoutTotalPrice}`}</p>
    </CheckoutSuccessfulModalWrapper>
  );
}