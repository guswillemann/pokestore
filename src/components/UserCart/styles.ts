import styled, { css } from 'styled-components';
import { getThemeColor } from '../../theme/utils';
import Button from '../Button';

const UserCartWrapper = styled.aside<{ isVisible: boolean }>`
  display: grid;
  grid-template-rows: auto 1fr 3rem 5rem;
  gap: 1rem;

  padding: 1rem;
  background-color: ${getThemeColor('secondary')};
  border: 2px solid ${getThemeColor('primary')};

  position: fixed;
  inset: 7.2rem 1rem 1rem 1rem;
  transition: opacity 500ms;
  max-width: 30rem;
  margin-left: auto;

  z-index: 100;

  @media(max-width: 1000px) {
    ${({ isVisible }) => !isVisible && css`
      opacity: 0;
      pointer-events: none;
    `}
  }
  
  border-radius: ${({ theme }) => theme.borderRadius};

  @media(min-width: 1000px) {
    right: 1rem;
    left: initial;
    width: 30rem;
  }

  h3 {
    text-align: center;
    color: white;
    font-size: 3rem;
  }

  ul {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    overflow-y: auto;
    overflow-x: hidden;
  }

  .total-price {
    color: white;
    font-weight: 700;
    font-size: 2rem;
  }
`;

export const CheckoutButton = styled(Button)`
  height: 100%;
`;

export default UserCartWrapper;
