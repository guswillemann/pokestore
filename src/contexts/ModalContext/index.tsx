import { createContext, ReactNode, useContext, useState } from 'react';
import LockScroll from '../../components/LockScroll';
import Modal from './Modal';

type ModalContextTypes = {
  modalContents: ReactNode;
  activeModal: (contents: ReactNode) => void;
  toggleModal: () => void;
}

export const ModalContext = createContext({} as ModalContextTypes);

type ModalProviderProps = {
  children: ReactNode;
}

export function ModalProvider({ children }: ModalProviderProps) {
  const [modalContents, setModalContents] = useState<ReactNode>(null);
  const [isModalVisible, setIsModalVisible] = useState(false);

  const hasModalContents = Boolean(modalContents);

  function activeModal(contents: ReactNode) {
    setModalContents(contents);
    setIsModalVisible(true);
  }
  
  function toggleModal() {
    setIsModalVisible(!isModalVisible);
  }

  return (
    <ModalContext.Provider value={{
      modalContents,
      activeModal,
      toggleModal,
    }}>
      {isModalVisible && <LockScroll />}
      {hasModalContents && <Modal isVisible={isModalVisible} />}
      {children}
    </ModalContext.Provider>
  );
}

export default function useModal() {
  return useContext(ModalContext);
};
