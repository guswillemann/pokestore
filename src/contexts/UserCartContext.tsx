import { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import nookies from 'nookies';
import { useRouter } from 'next/router';
import useToast from './ToastContext';

export type ProductType = {
  id: string;
  name: string;
  price: string;
  img: string;
}

type UserCartContextTypes = {
  cartList: Array<ProductType>;
  addToCart: (product: ProductType) => void;
  resetCart: () => void;
  toggleCart: () => void;
  removeItem: (index: number) => void;
  isCartVisible: boolean;
}

export const UserCartContext = createContext({} as UserCartContextTypes);

type UserCartProviderProps = {
  children: ReactNode;
  cartListCookie: string;
  avaibleProducts: Array<ProductType>;
}

export const CART_LIST_COOKIE = 'CART_LIST_COOKIE';

export function UserCartProvider({ children, cartListCookie, avaibleProducts }: UserCartProviderProps) {
  const { pathname } = useRouter();
  const { activeToast } = useToast();
  const [isCartVisible, setIsCartVisible] = useState(false);
  const [cartList, setCartList] = useState<Array<ProductType>>([]);

  useEffect(() => {
      if (cartListCookie) {
        const cookie = cartListCookie || nookies.get()[CART_LIST_COOKIE];
        const cookieCartItems = cookie.split(',');
        const initialCartList = avaibleProducts.filter((product) => cookieCartItems.some((cartItem) => {
          return product.id === cartItem;
        }));
        setCartList(initialCartList);
      }
  }, [cartListCookie])

  function addToCart(product: ProductType) {
    const hasProduct = cartList.some((cartItem) => product.id === cartItem.id)
    if (!hasProduct) {
      const newCartList = [...cartList, product];
      setCartList(newCartList);
      
      const cartIdList = newCartList.reduce((list, item) => [...list, item.id], [] as string[]);
      nookies.set(null, CART_LIST_COOKIE, cartIdList.toString(), {
        path: pathname,
        maxAge: 60 * 60 * 24 * 7,
      })
      return
    }

    activeToast({
      message: 'Você já possui este item no carrinho',
      type: 'alert',
    })
  }

  function resetCart() {
    setCartList([]);
    nookies.destroy(null, CART_LIST_COOKIE, { path: pathname });
  }

  function toggleCart() {
    setIsCartVisible(!isCartVisible);
  }

  function removeItem(index: number) {
    const newCartList = [...cartList];
    newCartList.splice(index, 1);
    setCartList(newCartList);
  }

  return (
    <UserCartContext.Provider value={{
      cartList,
      addToCart,
      resetCart,
      toggleCart,
      removeItem,
      isCartVisible,
    }}>
      {children}
    </UserCartContext.Provider>
  );
}

export default function useUserCart() {
  return useContext(UserCartContext);
};
