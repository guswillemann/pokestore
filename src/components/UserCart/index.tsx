import { useEffect, useState } from 'react';
import styled, { css } from 'styled-components';
import useModal from '../../contexts/ModalContext';
import useUserCart from '../../contexts/UserCartContext';
import { getThemeColor } from '../../theme/utils';
import LockScroll from '../LockScroll';
import CartItem from './CartItem';
import CheckoutSuccessfulModal from './CheckoutSuccessfulModal';

const UserCartWrapper = styled.aside<{ isVisible: boolean }>`
  display: grid;
  grid-template-rows: 2rem 1fr 3rem 5rem;
  gap: 1rem;

  padding: 1rem;
  background-color: ${getThemeColor('secondary')};
  border: 2px solid ${getThemeColor('primary')};

  position: fixed;
  inset: 6rem 1rem 1rem 1rem;
  transition: opacity 500ms;

  z-index: 100;

  @media(max-width: 1000px) {
    ${({ isVisible }) => !isVisible && css`
      opacity: 0;
      pointer-events: none;
    `}
  }
  
  border-radius: ${({ theme }) => theme.borderRadius};

  @media(min-width: 1000px) {
    right: 1rem;
    left: initial;
    width: 30rem;
  }

  h3 {
    text-align: center;
  }

  ul {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    overflow-y: auto;
    overflow-x: hidden;
  }
`;

const CheckoutButton = styled.button`
  border-radius: inherit;
  border: none;
  cursor: pointer;

  &:hover, &:focus {
    background-color: ${getThemeColor('primary')};
    color: white;
  }
`;

export default function UserCart() {
  const {
    cartList,
    resetCart,
    isCartVisible,
  } = useUserCart();
  const { activeModal } = useModal();

  const [priceSum, setPriceSum] = useState('0');

  function handleCheckout() {
    activeModal(<CheckoutSuccessfulModal CheckoutTotalPrice={priceSum} />)
    resetCart();
  }

  useEffect(() => {
    const newPriceSum = cartList.reduce((sum, item) => {
      return (Number(sum) + Number(item.price)).toFixed(2);
    }, '0');

    setPriceSum(newPriceSum);
  }, [cartList])

  return (
    <UserCartWrapper isVisible={isCartVisible}>
      {isCartVisible && <LockScroll />}
      <h3>Carrinho</h3>
      <ul>
        {cartList.map((product, index) => (
          <CartItem key={product.id} product={product} index={index} />
        ))}
      </ul>
      <p>Total: R$ {priceSum}</p>
      <CheckoutButton type="button" onClick={handleCheckout}>Finalizar</CheckoutButton>
    </UserCartWrapper>
  );
}