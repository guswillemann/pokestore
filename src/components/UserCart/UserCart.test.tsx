import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FC } from 'react';
import { ThemeProvider } from 'styled-components';
import UserCart from '.';
import { testTheme } from '../../../test/test-utils';
import { ModalContext } from '../../contexts/ModalContext';
import { ProductType, UserCartContext } from '../../contexts/UserCartContext';

const generateCartList = (itemsAmount: number) => {
  const cartList: Array<ProductType> = [];
  
  for (let i = 0;i<itemsAmount;i++) {
    cartList.push({
      id: i.toString(),
      name: 'testProduct',
      img: 'img',
      price: '20',
    })
  }

  return cartList;
}

const activeModalMock = jest.fn();
const resetCartMock = jest.fn();

function renderUserCart(cartList = generateCartList(1)) {
  const Wrapper: FC = ({ children }) => (
    <ThemeProvider theme={testTheme}>
      <ModalContext.Provider value={{
        activeModal: activeModalMock,
      } as any}>
        <UserCartContext.Provider value={{
          cartList,
          resetCart: resetCartMock,
          isCartVisible: true,
        } as any}>
          {children}
        </UserCartContext.Provider>
      </ModalContext.Provider>
    </ThemeProvider>
  );
  
  return render(<UserCart />, { wrapper: Wrapper });
}

describe('<UserCart />', () => {
  it('should match the snapshot', () => {
    const userCart = renderUserCart();
    expect(userCart).toMatchSnapshot();
  });

  describe('when the finish checkout button is clicked', () => {
    it('should active a modal and reset the cart', () =>{
      renderUserCart();
      const checkoutButton = screen.getByText('Checkout');
      
      expect(checkoutButton).toBeInTheDocument();
      userEvent.click(checkoutButton);
      
      expect(activeModalMock).toBeCalledTimes(1);
      expect(resetCartMock).toBeCalledTimes(1);
    });
  });

  describe('when the cart list is empty', () => {
    it('should not render any <CartItem />', () => {
      renderUserCart([]);

      const cartItems = screen.queryAllByTestId('cart-item');
      expect(cartItems.length).toEqual(0);
    });
  });

  describe('when the cart list has a single item', () => {
    it('should render a single <CartItem />', () => {
      renderUserCart();

      const cartItems = screen.getAllByTestId('cart-item');
      expect(cartItems.length).toEqual(1);
    });
  });
  
  describe('when the cart list has three items', () => {
    it('should render three <CartItem />', () => {
      renderUserCart(generateCartList(3));

      const cartItems = screen.getAllByTestId('cart-item');
      expect(cartItems.length).toEqual(3);
    });
  });
});
