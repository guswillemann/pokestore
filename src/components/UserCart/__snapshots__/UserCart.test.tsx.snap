// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserCart /> should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  -webkit-animation: slideLeft 500ms ease-in;
  animation: slideLeft 500ms ease-in;
  display: grid;
  gap: 1rem;
  grid-template-columns: 4rem 1fr 10rem 1rem;
  height: 5rem;
  background-color: #FAFAFA;
  padding: 0.5rem;
  border-radius: 8px;
}

.c1 p {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c1 .price-tag {
  margin-left: auto;
}

.c1 button {
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 0;
}

.c1 button:hover,
.c1 button:focus {
  opacity: 0.5;
}

.c0 {
  display: grid;
  grid-template-rows: 2rem 1fr 3rem 5rem;
  gap: 1rem;
  padding: 1rem;
  background-color: pink;
  border: 2px solid darkred;
  position: fixed;
  inset: 6rem 1rem 1rem 1rem;
  -webkit-transition: opacity 500ms;
  transition: opacity 500ms;
  z-index: 100;
  border-radius: 8px;
}

.c0 h3 {
  text-align: center;
}

.c0 ul {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  overflow-y: auto;
  overflow-x: hidden;
}

.c2 {
  border-radius: inherit;
  border: none;
  cursor: pointer;
}

.c2:hover,
.c2:focus {
  background-color: darkred;
  color: white;
}

@media (max-width:1000px) {

}

@media (min-width:1000px) {
  .c0 {
    right: 1rem;
    left: initial;
    width: 30rem;
  }
}

<body>
    <div>
      <aside
        class="c0"
      >
        <h3>
          Carrinho
        </h3>
        <ul>
          <li
            class="c1"
            data-testid="cart-item"
          >
            <img
              alt="Imagem do produto"
              src="img"
            />
            <p>
              testProduct
            </p>
            <p
              class="price-tag"
            >
              R$ 20
            </p>
            <button
              aria-label="Remover item do carrinho"
              type="button"
            >
              <img
                alt="X"
                src="/images/x-icon.svg"
              />
            </button>
          </li>
        </ul>
        <p>
          Total: R$ 
          20.00
        </p>
        <button
          class="c2"
          type="button"
        >
          Finalizar
        </button>
      </aside>
    </div>
  </body>,
  "container": <div>
    <aside
      class="UserCart__UserCartWrapper-sc-dz0mse-0 gmdhmN"
    >
      <h3>
        Carrinho
      </h3>
      <ul>
        <li
          class="CartItem__CartItemWrapper-sc-13akc3r-0 dMWAEM"
          data-testid="cart-item"
        >
          <img
            alt="Imagem do produto"
            src="img"
          />
          <p>
            testProduct
          </p>
          <p
            class="price-tag"
          >
            R$ 20
          </p>
          <button
            aria-label="Remover item do carrinho"
            type="button"
          >
            <img
              alt="X"
              src="/images/x-icon.svg"
            />
          </button>
        </li>
      </ul>
      <p>
        Total: R$ 
        20.00
      </p>
      <button
        class="UserCart__CheckoutButton-sc-dz0mse-1 jVchXm"
        type="button"
      >
        Finalizar
      </button>
    </aside>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
